common/core/desktop/src/kmx/kmx_capslock.cpp

```diff
@@ -2,57 +2,80 @@
   Copyright:        Copyright (C) 2003-2018 SIL International.
   Authors:          mcdurdin
 */
+#include <keyman/keyboardprocessor_consts.h>
 #include <kmx/kmx_processevent.h>
 
 using namespace km::kbp;
 using namespace kmx;
 
-void KMX_ProcessEvent::ResetCapsLock(void)
-{
-  DebugLog("ResetCapsLock: enter");
-
-  if (m_keyboard.Keyboard->dwFlags & KF_CAPSALWAYSOFF)
-  {
-    DebugLog("ResetCapsLock: caps lock should be always off");
-    if (m_environment.capsLock())
-    {
-      DebugLog("ResetCapsLock: caps lock is on, switching off caps lock");
-      m_actions.QueueAction(QIT_CAPSLOCK, 0);
-    }
+/**
+ * Set caps lock state.
+ * Called by ProcessEvent.
+ * May update `modifiers` to add or remove `CAPITALFLAG` according to capsLockOn
+ * parameter. May queue actions to set caps lock state.
+ *
+ * @param[in,out] modifiers   The modifier key bitmap
+ * @param         capsLockOn  The desired caps lock state
+ */
+void KMX_ProcessEvent::SetCapsLock(KMX_DWORD &modifiers, KMX_BOOL capsLockOn) {
+  KMX_BOOL capsLockCurrentlyOn = IsCapsLockOn(modifiers);
+  if (capsLockCurrentlyOn == capsLockOn) {
+    return;
```

Comment:
Should we assert that `capsLockOn` matches expected `modifiers` bitflag?

Reply:
What and where would you assert?
Maybe I miss the point, but this method gets called to bring CapsLock in the desired state, regardless of what it currently is.

Reply:
Yes, you are right. We calculate `capsLockCurrentlyOn` from `modifiers` so ignore my original question!